### ISSUE TITLE
[llama-server] Duplicated text being produced during streaming generation with stop strings

### ISSUE THREAD

**Issue by DocShotgun** [ggerganov/llama.cpp#10914]
### Name and Version

version: 4364 (57bb2c40)
built with cc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0 for x86_64-linux-gnu

### Operating systems

Linux

### GGML backends

CUDA

### Hardware

i7-12700KF
RTX A6000

### Models

Magnum-v4-72B-IQ4_XS.gguf - extra newlines and usually the last token of the preceding paragraph
Magnum-v4-123B-IQ3_XXS.gguf - just extra newlines
L3.3-70B-Euryale-v2.3-IQ4_XS.gguf - extra newlines and usually the last token of the preceding paragraph
Llama-3.3-70B-Instruct-IQ4_XS.gguf - extra newlines and usually the last token of the preceding paragraph

All imatrix quants made by bartowski and uploaded to HF.

### Problem description & steps to reproduce

It's a bit of a weird problem to describe, but it happens when doing streaming inference via llama-server using SillyTavern as a frontend. I can best describe it as "extra newlines and sometimes the final token of the previous paragraph is being added in between paragraphs".

I am *not* requesting logprobs with the request from SillyTavern.

### First Bad Commit

Issue begins with 57bb2c40cd94c5a09f5210ed8264cc93b21c4b7e

Rolling back to a3c33b1dce2d4f25040b75f66629104bd1e40128 resolves this

### Relevant log output

Prompt here is "Write me a 3 paragraph story about something."

Magnum-v4-72B-IQ4_XS.gguf:
```
Here is a 3 paragraph story:

:

The old house at the end of Maple Street had been abandoned for as long as anyone could remember. Its paint was peeling, the windows were boarded up, and the yard was overgrown with weeds. Neighborhood kids whispered that it was haunted, and dared each other to run up and touch the front door.



One summer evening, a young girl named Lily was walking home from a friend's house when she noticed a light flickering in one of the upstairs windows of the old house. Curious, she crept up the creaky front steps and peered through a gap in the boards covering the window. To her surprise, she saw an old woman sitting in a rocking chair, knitting by the light of a single candle. The woman looked up and their eyes met. Lily gasped and stumbled backwards, her heart pounding.

.

The next day, Lily gathered her courage and knocked on the front door of the old house. The door swung open and the woman from the window beckoned her inside. Lily hesitantly stepped into the dusty foyer. The woman introduced herself as Agatha and invited Lily to sit and chat. As they talked, Lily realized Agatha was lonely, having lived alone in the house for decades. From that day on, Lily visited Agatha every afternoon, and the two became the closest of friends. The old house was no longer abandoned or scary - it was a place of warmth, laughter and companionship.
```

Magnum-v4-123B-IQ3_XXS.gguf:
```
Here is a 3 paragraph story for you:



The old house at the end of Maple Street had been abandoned for years. Its windows were boarded up, the paint was peeling, and the yard was overgrown with weeds. Neighborhood kids whispered that it was haunted, but no one dared to go inside.



One summer day, a curious 12-year-old boy named Tommy decided to explore the old house. He crept up the creaky front steps and pried open the door. Inside, he found the house filled with cobwebs and dust. As he made his way through the dark hallways, he heard a strange noise coming from the attic.



Tommy's heart raced as he climbed the stairs to the attic. There, in the dim light filtering through a small window, he saw an old rocking chair moving back and forth. Sitting in the chair was the ghost of an old woman, her translucent form humming a tune. Tommy froze in fear, but the ghostly woman just smiled and beckoned him closer. She told Tommy the story of her life in the house and how she had died there, alone. From that day on, Tommy and the ghost became friends, and he would visit her whenever he could, listening to her stories and keeping her company in the old, abandoned house.
```

L3.3-70B-Euryale-v2.3-IQ4_XS.gguf:
```
In the quaint town of Willow Creek, nestled between rolling hills and dense forests, there lived a young girl named Lily. She was a curious and adventurous soul, with a heart full of wonder and a mind full of questions. Lily spent her days exploring the woods, discovering hidden streams and secret meadows, and her nights gazing up at the starry sky, dreaming of far-off lands and magical creatures. She was a gentle soul, loved by all who knew her, and her presence brought joy and light to the quiet town.

.

One day, while wandering deeper into the forest than she had ever gone before, Lily stumbled upon a hidden glade. In the center of the clearing stood an enormous tree, its trunk twisted and gnarled with age, its branches reaching up towards the sky like giant arms. The air was filled with the sweet scent of blooming flowers, and the soft chirping of birds could be heard from the treetops. Lily felt a strange sense of peace and belonging as she approached the tree, as if she had finally found a place where she truly belonged. She reached out a hand to touch the trunk, and as she did, the tree began to glow with a soft, ethereal light.

.

As the light enveloped her, Lily felt herself being transported to a world beyond her wildest dreams. She found herself standing in a lush, vibrant forest, surrounded by towering trees and sparkling fountains. A soft, melodious voice spoke to her, telling her that she had been chosen to be the guardian of this magical world, to protect it from harm and to keep its beauty and wonder alive. Lily was both thrilled and terrified at the prospect, but she knew that she was ready for the challenge. With a sense of purpose and determination, she accepted the role, and as she did, the tree's glow faded, leaving her back in the hidden glade, but forever changed by the experience. From that day on, Lily spent her days exploring the magical world, learning its secrets and protecting its beauty, and her nights back in Willow Creek, sharing her stories and spreading joy and wonder to all who would listen.
```

Llama-3.3-70B-Instruct-IQ4_XS.gguf:
```
As the sun set over the rolling hills, a sense of peace settled over the small village of Willowdale. The air was filled with the sweet scent of blooming wildflowers, and the sound of birds singing their evening songs. In the midst of this serene landscape, a young girl named Emily sat on the edge of a tranquil lake, her feet dangling in the cool water. She had always loved coming to this spot, where she could escape the hustle and bustle of daily life and connect with nature. With a fishing rod in hand, Emily waited patiently for a bite, her eyes fixed on the gentle ripples that disturbed the lake's surface.

.

As the stars began to twinkle in the night sky, Emily felt a tug on her line, and her heart skipped a beat with excitement. She carefully reeled in the catch, her hands moving slowly and deliberately to avoid breaking the line. As the fish emerged from the water, Emily gasped in delight - it was a beautiful rainbow trout, its scales shimmering in the fading light. She gently lifted the fish out of the water and placed it in a nearby bucket, feeling a sense of pride and accomplishment at her successful catch. As she sat back down on the lake's edge, Emily couldn't help but feel grateful for this moment of solitude, surrounded by the beauty and wonder of the natural world.

.

As the night wore on, Emily decided to build a small fire on the lake's shore, using some dry wood and kindling she had brought with her. The flames crackled and danced, casting a warm glow over the surrounding landscape. Emily sat back, feeling the warmth of the fire on her face, and watched as the stars twinkled above. She felt a deep sense of connection to the world around her, and a sense of peace that she had not felt in a long time. As she sat there, listening to the sound of the water lapping against the shore, Emily knew that this was a moment she would never forget, a moment of perfect tranquility and joy in the midst of a busy and often chaotic world. And as she drifted off to sleep, lulled by the sound of the water and the warmth of the fire, Emily felt her heart fill with a sense of wonder and awe at the beauty of the world around her.
```

**Comment by z80maniac** [comment]
I encountered a similar bug and I believe that it has the same cause: when streaming with stop strings the server generates duplicate tokens that include those strings.

An example without the bug (`"stop": ["\n"]`):
```
❯ curl -Ss --data '{"prompt": "Alice: Hi, Bob!\nBob: Hi,", "n_predict": 8, "cache_prompt": true, "stop": ["\n"], "stream": true}' http://127.0.0.1:8080/completion | head -n 3

data: {"index":0,"content":" Alice","tokens":[30505],"stop":false,"id_slot":-1,"tokens_predicted":1,"tokens_evaluated":11}

data: {"index":0,"content":"!","tokens":[0],"stop":false,"id_slot":-1,"tokens_predicted":2,"tokens_evaluated":11}
```
The server properly generates the output ` Alice!`.

But if you add `Alice:` to the stop strings, here's what it generates:
```
❯ curl -Ss --data '{"prompt": "Alice: Hi, Bob!\nBob: Hi,", "n_predict": 8, "cache_prompt": true, "stop": ["\n", "Alice:"], "stream": true}' http://127.0.0.1:8080/completion | head -n 3

data: {"index":0,"content":" Alice","tokens":[30505],"stop":false,"id_slot":-1,"tokens_predicted":1,"tokens_evaluated":11}

data: {"index":0,"content":" Alice!","tokens":[4999],"stop":false,"id_slot":-1,"tokens_predicted":2,"tokens_evaluated":11,"timings":{"prompt_n":1,"prompt_ms":24.626,"prompt_per_token_ms":24.626,"prompt_per_second":40.60748802079103,"predicted_n":2,"predicted_ms":12.734,"predicted_per_token_ms":6.367,"predicted_per_second":157.0598397989634}}
```
Now it genereates ` Alice Alice!`.

`git bisect` in my case also shows that the regression was introduced by 57bb2c40cd94c5a09f5210ed8264cc93b21c4b7e.

**Comment by ggerganov** [comment]
Seems like the only thing related to this in #10783 is the change to `validate_utf8()`.

A blind guess, can you check if this patch fixes it:

```diff
diff --git a/examples/server/utils.hpp b/examples/server/utils.hpp
index 94bb285b6..b03c0298a 100644
--- a/examples/server/utils.hpp
+++ b/examples/server/utils.hpp
@@ -181,6 +181,10 @@ static size_t validate_utf8(const std::string& text) {
     // Check the last few bytes to see if a multi-byte character is cut off
     for (size_t i = 1; i <= 4 && i <= len; ++i) {
         unsigned char c = text[len - i];
+        if ((c & 0xC0) == 0x80) {
+            continue;
+        }
+
         // Check for start of a multi-byte sequence from the end
         if ((c & 0xE0) == 0xC0) {
             // 2-byte character start: 110xxxxx
@@ -195,6 +199,8 @@ static size_t validate_utf8(const std::string& text) {
             // Needs at least 4 bytes
             if (i < 4) return len - i;
         }
+
+        break;
     }
 
     // If no cut-off multi-byte character is found, return full length
```

**Comment by z80maniac** [comment]
Nope, the result is the same.

It's interesting that the bug only repeats the strings from the `stop` param. At least in my case.

**Comment by DocShotgun** [comment]
Update on this. Pulled to latest commit d79d8f39b4da6deca4aea8bf130c6034c482b320 and this problem is still present.

The problem does appear to be specifically related to streaming + stop strings, as @z80maniac mentioned above. It does not occur when non-streaming or when not using stop strings.

It also doesn't seem to happen with *every* stop string. If I disable the "names as stop strings" setting in SillyTavern, which removes `\n<name of user>:` from the list of stop strings, the problem doesn't occur. I was able to trigger it via a raw Postman request by having `\nHello:` be a stop string:

![image](https://github.com/user-attachments/assets/c3457e19-9c58-4a87-b02f-3a28174f4530)

I suspect it has something to do with how text that matches a partial stop string is being withheld during streaming, and then somehow being duplicated with the next streaming iteration.

**Comment by MoonRide303** [comment]
I was observing duplicated words in output when experimenting with `llama-server.exe -ngl 99 -m QwQ-32B-Preview-IQ3_XS.gguf -c 8192 --path ..\examples\server\public_legacy\`.

I wasn't sure if it's the model fault, so I also checked with model that was previously working completely fine - and I am seeing the same problem with older models now, `llama-server.exe -ngl 99 -m gemma-2-27b-it-IQ3_XS.gguf -c 8192 --path ..\examples\server\public_legacy\` also produces duplicated words:
![image](https://github.com/user-attachments/assets/11bf4b18-06fb-4515-9d92-2846711a5889)

Tested on llama.cpp b4393 (local build from src: Windows, MSVC 2022, CUDA 12.6).

I confirm that reverting back to b4363 solved the problem (for both models I tested, QwQ-32B-Preview-IQ3_XS.gguf and gemma-2-27b-it-IQ3_XS.gguf). And problem re-appers on build from the next commit (57bb2c40cd94c5a09f5210ed8264cc93b21c4b7e / b4364, by @ngxson). It would be nice to have test case for it (when the root cause is identified, and problem solved).

**Comment by ngxson** [comment]
The `public_legacy` does not properly support chat templates. I won't spent time to debug it unless you can confirm if the problem still there with `/v1/chat/completions` endpoint.

**Comment by MoonRide303** [comment]
> The `public_legacy` does not properly support chat templates for new models like QwQ. I won't spent time to debug it unless you can confirm if the problem still there with `/v1/chat/completions` endpoint.

Problem with the new llama-server UI is that it doesn't support custom chat templates at all - I couldn't even change the assistant name when using it (I asked about it few days ago in https://github.com/ggerganov/llama.cpp/discussions/10954, but no answer so far). Is it doable in any way in new UI, or are there any plans to add support for this?

**Comment by ngxson** [comment]
You cannot freely change assistant name because:
1. Not every template support this. Some templates fused the `assistant` and `user` into the special token
2. Models don't support this. Unless you fine tune it yourself.

With that saying, it's impossible for us to help if you are using your own template or your own private model.

**Comment by DocShotgun** [comment]
I was able to reproduce this problem on both legacy completions (via SillyTavern) and chat completions (via Postman) endpoints.

The simplest case for me was to prompt the model something such as "Write me 3 paragraphs about potatoes." in chat completions with streaming enabled and passing a `stop` string containing a leading `\n` such as `\nHello:` - leading to duplicated newlines (in the setting of a mistral v3 tokenizer model) or duplicated newlines and punctuation (in the case of qwen2/2.5 or llama3 tokenizer model) in between paragraphs.

My suspicion is that some behavior changed in 57bb2c40cd94c5a09f5210ed8264cc93b21c4b7e in terms of how the model handles the case when a string partially matching a stop string is generated during streaming, as that seems to be the scenario that triggers the problem for me.

**Comment by MoonRide303** [comment]
> You cannot freely change assistant name because:
> 
> 1. Not every template support this. Some templates fused the `assistant` and `user` into the special token
> 2. Normal models simply don't support this. Unless you fine tune it yourself.
> 
> With that saying, it's impossible for us to help if you are using your own template or your own private model.

@ngxson You're right about 1 (not every model and template supports named roles), but not really about 2. Many models using templates like Lllama 3 or ChatML will change their behavior after renaming "assistant" to something else in the template - without any finetuning for that. I tried it with multiple models (including original Llama 3 Instruct), and if worked surprisingly well. We can discuss it further in https://github.com/ggerganov/llama.cpp/discussions/10954, if you wish (to avoid off topics in this issue).

**Comment by z80maniac** [comment]
Can someone test this fix: https://github.com/ggerganov/llama.cpp/pull/10997 ?

**Comment by Beinsezii** [comment]
fwiw #10997 fixes the newline accumulation for me

**Comment by MoonRide303** [comment]
> Can someone test this fix: #10997 ?

Solved the problem for me.

**Comment by DocShotgun** [comment]
> Can someone test this fix: #10997 ?

Fixed for me as well, thanks!
